### PR TITLE
feature(planner): Translate subquery into apply operator

### PR DESCRIPTION
**Pull request by leiysky**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Translate subqueries into apply operators, so we can represent all the subqueries(correlated or uncorrelated) with relational algebra.

Besides, this PR also refines the property derivation mechanism with `RelExpr`.

## Changelog

- New Feature

## Related Issues

Part of #5210

### CODE REVIEW

File: query/src/sql/planner/mod.rs [hunk 1]
```diff
@@ -13,6 +13,7 @@
 // limitations under the License.
 
 use std::sync::Arc;
+use std::sync::RwLock;
```
Comment (BohuTANG) [inline query/src/sql/planner/mod.rs:16]: It's better to use the common/base/core/infallible rwlock https://github.com/datafuselabs/databend/blob/main/common/base/src/infallible/rwlock.rs , so we no need do any unwrap.

Comment (leiysky) [reply query/src/sql/planner/mod.rs:16]: That's good. 

By the way, I'd like to know if it is the canonical `Mutex` of databend?

Comment (BohuTANG) [reply query/src/sql/planner/mod.rs:16]: Infallible rwlock is a simple wrapper around the lock() function of a std::sync::RwLock The only difference is that you don’t need to call unwrap() on it.

Infallible Mutex https://github.com/datafuselabs/databend/blob/main/common/base/src/infallible/mutex.rs

Comment (leiysky) [reply query/src/sql/planner/mod.rs]: Fixed.

**Review by BohuTANG — APPROVED** [review]
:100:

**Comment by BohuTANG** [comment]
Let's merge it first, the review remains open.

**Pull request merged**